src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -729,57 +758,54 @@ def _infer_sharing_attributes(self):
                     last_read_position = access.node.abs_position
 
                 if access.access_type == AccessType.WRITE:
-                    # Check if the write access is inside a loop. If the write
-                    # is outside of a loop, it is an assignment to a shared
-                    # variable. Example where jpk is likely used outside of the
-                    # parallel section later, so it must be declared as shared
-                    # in order keep its value:
+
+                    # Check if the write access is outside a loop. In this case
+                    # it will be marked as shared. This is done because it is
+                    # likely to be re-used later. e.g:
                     # !$omp parallel
                     # jpk = 100
                     # !omp do
                     # do ji = 1, jpk
-
-                    # TODO #598: Improve the handling of scalar variables,
-                    # there are remaining issues with references after the
-                    # parallel region of variables that we currently declare
-                    # as private.
-
-                    # Check if it is inside a loop
                     loop_ancestor = access.node.ancestor(
                         (Loop, WhileLoop),
                         limit=self,
                         include_self=True)
+                    if not loop_ancestor:
+                        # If we find it at least once outside a loop we keep it
+                        # as shared
+                        break
 
-                    if loop_ancestor:
-                        # The assignment to the variable is inside a loop, so
-                        # declare it to be private
-                        name = str(signature).lower()
-                        symbol = access.node.scope.symbol_table.lookup(name)
-
-                        # Is the write conditional?
-                        conditional_write = access.node.ancestor(
-                            IfBlock,
-                            limit=loop_ancestor,
-                            include_self=True)
-
-                        if has_been_read:
-                            loop_pos = loop_ancestor.loop_body.abs_position
-                            if last_read_position > loop_pos:
-                                # It is read in the same loop_body, so it is
-                                # a race condition
-                                need_sync.add(symbol)
-                            else:
-                                fprivate.add(symbol)
-                        elif conditional_write:
+                    # Otherwise, the assignment to this variable is inside a
+                    # loop (and it will be repeated for each iteration), so
+                    # we declare it as private or need_synch
+                    name = str(signature).lower()
```

Comment:
What happens if the variable in question is some sort of structure access, e.g. `a%b`? Will this still work?

Reply:
Just use `signature[0]` or `signature.var_name` (if the latter doesn't exceed line length :)  ). I don't think lower is required, symbol table `_normalize`s the name it looks up, which means it converts it to lower case.

Reply:
I used the suggested line. Although this just considers the top-level component of the derived type accessor. I added a TODO for the more complex types.